File: C++/2-Medium/00093._复原_IP_地址/00093._复原_IP_地址.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

class Solution{
public:
    vector<string> restoreIpAddresses(string s){
        vector<string> result;
        vector<string> ip;
        backtrack(s, 0, ip, result);
        return result;
    }

    void backtrack(string s, int start, vector<string>& ip, vector<string>& result){
        // 如果已经得到了四段ip地址并且start已经到了s的末尾
        if(ip.size() == 4 && start == s.size()){
            result.push_back(ip[0] + '.' + ip[1] + '.' + ip[2] + '.' + ip[3]);
            return;
        }

        if(ip.size() == 4 || start == s.size()){
            return;
        }

        for(int len = 1; len <= 3; len++){
            if(start + len > s.size()){
                break;
            }

            string segment = s.substr(start, len);
            if(segment.size() > 1 && segment[0] == '0'){
                break;
            }
            if(stoi(segment) > 255){
                break;
            }
            ip.push_back(segment);
            backtrack(s, start + len, ip, result);
            ip.pop_back();
        }
    }
};

int main() {
    Solution solution;
    string s = "25525511135";
    vector<string> result = solution.restoreIpAddresses(s);
    for(string ip : result) {
        cout << ip << endl;
    }
    return 0;
}

File: Python/2-Medium/00093._复原_IP_地址/00093._复原_IP_地址.py
from typing import List

class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        # 存储结果的列表
        result = []
        
        # 递归搜索函数
        def backtrack(start, path):
            # 达到四个整数且字符串已经遍历完
            if start == len(s) and len(path) == 4:
                result.append('.'.join(path))
                return
            
            # 尝试分割下一个整数
            for i in range(1, 4):
                # 分割的整数不越界
                if start + i <= len(s):
                    # 取出当前整数
                    num_str = s[start:start+i]
                    # 判断整数是否在有效范围
                    if 0 <= int(num_str) <= 255 and (num_str[0] != '0' or len(num_str) == 1):
                        # 递归搜索下一个整数
                        backtrack(start + i, path + [num_str])
        
        # 调用递归函数
        backtrack(0, [])
        return result

# 测试代码
solution = Solution()
result = solution.restoreIpAddresses("25525511135")
print(result)